File: src/cpp/Source/Chunks/Stat/StatisticsChunk.cpp
#include "PCH.h"
#include "StatisticsChunk.h"

using namespace std;
using namespace SlimShader;

shared_ptr<StatisticsChunk> StatisticsChunk::Parse(BytecodeReader& reader, const uint32_t chunkSize)
{
	auto size = chunkSize / sizeof(uint32_t);

	// Unknowns:
	// DefCount
	// MacroInstructionCount

	// PatchConstantParameters
	// cGSInstanceCount
	// cBarrierInstructions
	// cInterlockedInstructions
	// cTextureStoreInstructions

	auto result = shared_ptr<StatisticsChunk>(new StatisticsChunk());

	result->_instructionCount = reader.ReadUInt32();
	result->_tempRegisterCount = reader.ReadUInt32();
	result->_defineCount = reader.ReadUInt32(); // Guessed
	result->_declarationCount = reader.ReadUInt32();
	result->_floatInstructionCount = reader.ReadUInt32();
	result->_intInstructionCount = reader.ReadUInt32();
	result->_uintInstructionCount = reader.ReadUInt32();
	result->_staticFlowControlCount = reader.ReadUInt32();
	result->_dynamicFlowControlCount = reader.ReadUInt32();
	result->_macroInstructionCount = reader.ReadUInt32(); // Guessed
	result->_tempArrayCount = reader.ReadUInt32();
	result->_arrayInstructionCount = reader.ReadUInt32();
	result->_cutInstructionCount = reader.ReadUInt32();
	result->_emitInstructionCount = reader.ReadUInt32();
	result->_textureNormalInstructions = reader.ReadUInt32();
	result->_textureLoadInstructions = reader.ReadUInt32();
	result->_textureCompInstructions = reader.ReadUInt32();
	result->_textureBiasInstructions = reader.ReadUInt32();
	result->_textureGradientInstructions = reader.ReadUInt32();
	result->_movInstructionCount = reader.ReadUInt32();
	result->_movCInstructionCount = reader.ReadUInt32(); // Guessed
	result->_conversionInstructionCount = reader.ReadUInt32();
	result->_unknown4 = reader.ReadUInt32();
	result->_inputPrimitive = static_cast<Primitive>(reader.ReadUInt32());
	result->_geometryShaderOutputTopology = static_cast<PrimitiveTopology>(reader.ReadUInt32());
	result->_geometryShaderMaxOutputVertexCount = reader.ReadUInt32();
	result->_unknown5 = reader.ReadUInt32();
	result->_unknown6 = reader.ReadUInt32();
	result->_unknown7 = reader.ReadUInt32();

	// DX10 stat size
	if (size == 29)
		return result;

	// Unknown.
	assert(reader.ReadUInt32() == 0); // TODO

	result->_controlPoints = reader.ReadUInt32();
	result->_hullShaderOutputPrimitive = static_cast<TessellatorOutputPrimitive>(reader.ReadUInt32());
	result->_hullShaderPartitioning = static_cast<TessellatorPartitioning>(reader.ReadUInt32());
	result->_tessellatorDomain = static_cast<TessellatorDomain>(reader.ReadUInt32());

	result->_barrierInstructions = reader.ReadUInt32(); // Guessed
	result->_interlockedInstructions = reader.ReadUInt32(); // Guessed
	result->_textureStoreInstructions = reader.ReadUInt32(); // Guessed

	// DX11 stat size.
	if (size == 37)
		return result;

	throw runtime_error("Unhandled stat size: " + chunkSize);
}

StatisticsChunk::StatisticsChunk()
	: _controlPoints(0),
	_hullShaderOutputPrimitive(TessellatorOutputPrimitive::Undefined),
	_hullShaderPartitioning(TessellatorPartitioning::Undefined),
	_tessellatorDomain(TessellatorDomain::Undefined),
	_barrierInstructions(0),
	_interlockedInstructions(0),
	_textureStoreInstructions(0)
{

}

uint32_t StatisticsChunk::GetInstructionCount() const { return _instructionCount; }
uint32_t StatisticsChunk::GetTempRegisterCount() const { return _tempRegisterCount; }
uint32_t StatisticsChunk::GetTempArrayCount() const { return _tempArrayCount; }
uint32_t StatisticsChunk::GetDefineCount() const { return _defineCount; }
uint32_t StatisticsChunk::GetDeclarationCount() const { return _declarationCount; }
uint32_t StatisticsChunk::GetTextureNormalInstructions() const { return _textureNormalInstructions; }
uint32_t StatisticsChunk::GetTextureLoadInstructions() const { return _textureLoadInstructions; }
uint32_t StatisticsChunk::GetTextureCompInstructions() const { return _textureCompInstructions; }
uint32_t StatisticsChunk::GetTextureBiasInstructions() const { return _textureBiasInstructions; }
uint32_t StatisticsChunk::GetTextureGradientInstructions() const { return _textureGradientInstructions; }
uint32_t StatisticsChunk::GetFloatInstructionCount() const { return _floatInstructionCount; }
uint32_t StatisticsChunk::GetIntInstructionCount() const { return _intInstructionCount; }
uint32_t StatisticsChunk::GetUIntInstructionCount() const { return _uintInstructionCount; }
uint32_t StatisticsChunk::GetStaticFlowControlCount() const { return _staticFlowControlCount; }
uint32_t StatisticsChunk::GetDynamicFlowControlCount() const { return _dynamicFlowControlCount; }
uint32_t StatisticsChunk::GetMacroInstructionCount() const { return _macroInstructionCount; }
uint32_t StatisticsChunk::GetArrayInstructionCount() const { return _arrayInstructionCount; }
uint32_t StatisticsChunk::GetCutInstructionCount() const { return _cutInstructionCount; }
uint32_t StatisticsChunk::GetEmitInstructionCount() const { return _emitInstructionCount; }
PrimitiveTopology StatisticsChunk::GetGeometryShaderOutputTopology() const { return _geometryShaderOutputTopology; }
uint32_t StatisticsChunk::GetGeometryShaderMaxOutputVertexCount() const { return _geometryShaderMaxOutputVertexCount; }
Primitive StatisticsChunk::GetInputPrimitive() const { return _inputPrimitive; }
uint32_t StatisticsChunk::GetGeometryShaderInstanceCount() const { return _geometryShaderInstanceCount; }
uint32_t StatisticsChunk::GetControlPoints() const { return _controlPoints; }
TessellatorOutputPrimitive StatisticsChunk::GetHullShaderOutputPrimitive() const { return _hullShaderOutputPrimitive; }
TessellatorPartitioning StatisticsChunk::GetHullShaderPartitioning() const { return _hullShaderPartitioning; }
TessellatorDomain StatisticsChunk::GetTessellatorDomain() const { return _tessellatorDomain; }
uint32_t StatisticsChunk::GetBarrierInstructions() const { return _barrierInstructions; }
uint32_t StatisticsChunk::GetInterlockedInstructions() const { return _interlockedInstructions; }
uint32_t StatisticsChunk::GetTextureStoreInstructions() const { return _textureStoreInstructions; }
uint32_t StatisticsChunk::GetMovInstructionCount() const { return _movInstructionCount; }
uint32_t StatisticsChunk::GetMovCInstructionCount() const { return _movCInstructionCount; }
uint32_t StatisticsChunk::GetUnknown3() const { return _unknown3; }
uint32_t StatisticsChunk::GetConversionInstructionCount() const { return _conversionInstructionCount; }
uint32_t StatisticsChunk::GetUnknown4() const { return _unknown4; }
uint32_t StatisticsChunk::GetUnknown5() const { return _unknown5; }
uint32_t StatisticsChunk::GetUnknown6() const { return _unknown6; }
uint32_t StatisticsChunk::GetUnknown7() const { return _unknown7; }

std::ostream& SlimShader::operator<<(std::ostream& out, const StatisticsChunk& chunk)
{
	if (chunk._tessellatorDomain != TessellatorDomain::Undefined)
	{
		out << "// Tessellation Domain   # of control points" << endl;
		out << "// -------------------- --------------------" << endl;
		out << boost::format("// %-20s %20i") 
			% ToStringStat(chunk._tessellatorDomain)
			% chunk._controlPoints
			<< endl;
		out << "//" << endl;
	}
	if (chunk._hullShaderOutputPrimitive != TessellatorOutputPrimitive::Undefined)
	{
		out << "// Tessellation Output Primitive  Partitioning Type " << endl;
		out << "// ------------------------------ ------------------" << endl;
		out << boost::format("// %-30s %-18s")
			% ToStringStat(chunk._hullShaderOutputPrimitive)
			% ToStringStat(chunk._hullShaderPartitioning)
			<< endl;
		out << "//" << endl;
	}
	return out;
}

File: src/SlimShader/Chunks/Stat/StatisticsChunk.cs
using System.Diagnostics;
using System.Text;
using SlimShader.Chunks.Common;
using SlimShader.Util;

namespace SlimShader.Chunks.Stat
{
	/// <summary>
	/// Statistics chunk
	/// Based on D3D11_SHADER_DESC.
	/// </summary>
	public class StatisticsChunk : BytecodeChunk
	{
		/// <summary>
		/// The number of intermediate-language instructions in the compiled shader.
		/// </summary>
		public uint InstructionCount { get; private set; }

		/// <summary>
		/// The number of temporary registers in the compiled shader.
		/// </summary>
		public uint TempRegisterCount { get; private set; }

		/// <summary>
		/// Number of temporary arrays used.
		/// </summary>
		public uint TempArrayCount { get; private set; }

		/// <summary>
		/// Number of constant defines.
		/// </summary>
		public uint DefineCount { get; private set; }

		/// <summary>
		/// Number of declarations (input + output).
		/// </summary>
		public uint DeclarationCount { get; private set; }

		/// <summary>
		/// Number of non-categorized texture instructions.
		/// </summary>
		public uint TextureNormalInstructions { get; private set; }

		/// <summary>
		/// Number of texture load instructions
		/// </summary>
		public uint TextureLoadInstructions { get; private set; }

		/// <summary>
		/// Number of texture comparison instructions
		/// </summary>
		public uint TextureCompInstructions { get; private set; }

		/// <summary>
		/// Number of texture bias instructions
		/// </summary>
		public uint TextureBiasInstructions { get; private set; }

		/// <summary>
		/// Number of texture gradient instructions.
		/// </summary>
		public uint TextureGradientInstructions { get; private set; }

		/// <summary>
		/// Number of floating point arithmetic instructions used.
		/// </summary>
		public uint FloatInstructionCount { get; private set; }

		/// <summary>
		/// Number of signed integer arithmetic instructions used.
		/// </summary>
		public uint IntInstructionCount { get; private set; }

		/// <summary>
		/// Number of unsigned integer arithmetic instructions used.
		/// </summary>
		public uint UIntInstructionCount { get; private set; }

		/// <summary>
		/// Number of static flow control instructions used.
		/// </summary>
		public uint StaticFlowControlCount { get; private set; }

		/// <summary>
		/// Number of dynamic flow control instructions used.
		/// </summary>
		public uint DynamicFlowControlCount { get; private set; }

		/// <summary>
		/// Number of macro instructions used.
		/// </summary>
		public uint MacroInstructionCount { get; private set; }

		/// <summary>
		/// Number of array instructions used.
		/// </summary>
		public uint ArrayInstructionCount { get; private set; }

		/// <summary>
		/// Number of cut instructions used.
		/// </summary>
		public uint CutInstructionCount { get; private set; }

		/// <summary>
		/// Number of emit instructions used.
		/// </summary>
		public uint EmitInstructionCount { get; private set; }

		/// <summary>
		/// The <see cref="PrimitiveTopology" />-typed value that represents the geometry shader output topology.
		/// </summary>
		public PrimitiveTopology GeometryShaderOutputTopology { get; private set; }

		/// <summary>
		/// Geometry shader maximum output vertex count.
		/// </summary>
		public uint GeometryShaderMaxOutputVertexCount { get; private set; }

		/// <summary>
		/// Indicates whether a shader is a sample frequency shader.
		/// </summary>
		public bool IsSampleFrequencyShader { get; private set; }

		/// <summary>
		/// The <see cref="Primitive"/>-typed value that represents the input primitive for a geometry shader or hull shader.
		/// </summary>
		public Primitive InputPrimitive { get; private set; }

		/// <summary>
		/// Number of geometry shader instances.
		/// </summary>
		public uint GeometryShaderInstanceCount { get; private set; }

		/// <summary>
		/// Number of control points in the hull shader and domain shader.
		/// </summary>
		public uint ControlPoints { get; private set; }

		/// <summary>
		/// The <see cref="TessellatorOutputPrimitive"/>-typed value that represents the tessellator output-primitive type.
		/// </summary>
		public TessellatorOutputPrimitive HullShaderOutputPrimitive { get; private set; }

		/// <summary>
		/// The <see cref="TessellatorPartitioning" />-typed value that represents the tessellator partitioning mode.
		/// </summary>
		public TessellatorPartitioning HullShaderPartitioning { get; private set; }

		/// <summary>
		/// The <see cref="TessellatorDomain"/>-typed value that represents the tessellator domain.
		/// </summary>
		public TessellatorDomain TessellatorDomain { get; private set; }

		/// <summary>
		/// Number of barrier instructions in a compute shader.
		/// </summary>
		public uint BarrierInstructions { get; private set; }

		/// <summary>
		/// Number of interlocked instructions in a compute shader.
		/// </summary>
		public uint InterlockedInstructions { get; private set; }

		/// <summary>
		/// Number of texture writes in a compute shader.
		/// </summary>
		public uint TextureStoreInstructions { get; private set; }

		public uint MovInstructionCount { get; private set; }
		public uint MovCInstructionCount { get; private set; }
		public uint ConversionInstructionCount { get; private set; }

		public static StatisticsChunk Parse(BytecodeReader reader, uint chunkSize)
		{
			var size = chunkSize / sizeof(uint);

			var result = new StatisticsChunk
			{
				InstructionCount = reader.ReadUInt32(),
				TempRegisterCount = reader.ReadUInt32(),
				DefineCount = reader.ReadUInt32(),
				DeclarationCount = reader.ReadUInt32(),
				FloatInstructionCount = reader.ReadUInt32(),
				IntInstructionCount = reader.ReadUInt32(),
				UIntInstructionCount = reader.ReadUInt32(),
				StaticFlowControlCount = reader.ReadUInt32(),
				DynamicFlowControlCount = reader.ReadUInt32(),
				MacroInstructionCount = reader.ReadUInt32(), // Guessed
				TempArrayCount = reader.ReadUInt32(),
				ArrayInstructionCount = reader.ReadUInt32(),
				CutInstructionCount = reader.ReadUInt32(),
				EmitInstructionCount = reader.ReadUInt32(),
				TextureNormalInstructions = reader.ReadUInt32(),
				TextureLoadInstructions = reader.ReadUInt32(),
				TextureCompInstructions = reader.ReadUInt32(),
				TextureBiasInstructions = reader.ReadUInt32(),
				TextureGradientInstructions = reader.ReadUInt32(),
				MovInstructionCount = reader.ReadUInt32(),
				MovCInstructionCount = reader.ReadUInt32(),
				ConversionInstructionCount = reader.ReadUInt32()
			};

			var unknown0 = reader.ReadUInt32();
			Debug.Assert(unknown0 == 0);

			result.InputPrimitive = (Primitive) reader.ReadUInt32();
			result.GeometryShaderOutputTopology = (PrimitiveTopology) reader.ReadUInt32();
			result.GeometryShaderMaxOutputVertexCount = reader.ReadUInt32();

			var unknown1 = reader.ReadUInt32();
			Debug.Assert(unknown1 == 0 || unknown1 == 1);

			var unknown2 = reader.ReadUInt32();
			Debug.Assert(unknown2 == 0 || unknown2 == 2); // TODO

			result.IsSampleFrequencyShader = (reader.ReadUInt32() == 1);

			// DX10 stat size
			if (size == 29)
				return result;

			// Unknown.
			var unknown = reader.ReadUInt32();
			Debug.Assert(unknown == 0); // TODO

			result.ControlPoints = reader.ReadUInt32();
			result.HullShaderOutputPrimitive = (TessellatorOutputPrimitive) reader.ReadUInt32();
			result.HullShaderPartitioning = (TessellatorPartitioning) reader.ReadUInt32();
			result.TessellatorDomain = (TessellatorDomain) reader.ReadUInt32();

			result.BarrierInstructions = reader.ReadUInt32();
			result.InterlockedInstructions = reader.ReadUInt32();
			result.TextureStoreInstructions = reader.ReadUInt32();

			// DX11 stat size.
			if (size == 37)
				return result;

			throw new ParseException("Unhandled stat size: " + chunkSize);
		}

		public override string ToString()
		{
			var sb = new StringBuilder();
			if (TessellatorDomain != TessellatorDomain.Undefined)
			{
				sb.AppendLine("// Tessellation Domain   # of control points");
				sb.AppendLine("// -------------------- --------------------");
				sb.AppendLine(string.Format("// {0,-20} {1,20}", TessellatorDomain.GetDescription(),
					ControlPoints));
				sb.AppendLine("//");
			}
			if (HullShaderOutputPrimitive != TessellatorOutputPrimitive.Undefined)
			{
				sb.AppendLine("// Tessellation Output Primitive  Partitioning Type ");
				sb.AppendLine("// ------------------------------ ------------------");
				sb.AppendLine(string.Format("// {0,-30} {1,-18}", 
					HullShaderOutputPrimitive.GetDescription(ChunkType.Stat),
					HullShaderPartitioning.GetDescription(ChunkType.Stat)));
				sb.AppendLine("//");
			}
			if (IsSampleFrequencyShader)
			{
				sb.AppendLine("// Pixel Shader runs at sample frequency");
				sb.AppendLine("//");
			}
			return sb.ToString();
		}
	}
}